### ISSUE TITLE
Destructuring assignment makes any types when it skips some elements (regression)

### ISSUE THREAD

**Issue by falsandtru** [Microsoft/TypeScript#20872]
cc @andy-ms 

**TypeScript Version:**  2.7.0-dev.20171223

**Code**

```ts
const [, a = ''] = ''.match('') || [];
```

**Expected behavior:**

`a` is string.

**Actual behavior:**

`a` is any.

**Comment by weswigham** [comment]
@ahejlsberg The bisect indicates that 25af351014236059030ed4bc6b1c8aec2eee8978 introduced this regression as part of #19671. I think the change is unintentional, in this instance? Before, when we saw `RegExpMatchArray` on the LHS of `||` and `[any, ""]` on the RHS, we did a best-choice thing that eliminated the tuple type from the union. Now that it is preserved, when we go to get the index type for the union, we see a tuple with an `any` member (created from the shape of the destructuring pattern) and happily state that the numeric index type must thusly be `any`.  I think a decent fix that doesn't revert our behavior back to the best-choice mechanic is to stop using `unknownType` for positions indicated by `OmittedExpressions` when building the tuple type in the first place in `checkArrayLiteral` (since those broaden the index type of the tuple type created to `any`), and instead use `emptyObjectType` or `neverType`.
